### PR TITLE
allow validators-external-signer-public-keys to accept a url

### LINKED ISSUE #3378: Allow external signer public keys to be loaded from url
currently, `--validators-external-signer-public-keys` allows us to send in public keys used by an external signer, so that we know the public keys to use for validators.

It'd be nicer if they could be loaded from URL rather than having to get the public keys and then pass them into teku.

i would like to be able to do something like

`--validators-external-signer-public-keys-url http://localhost:9000/api/v1/eth2/publicKeys` 

and have teku load the public keys from http request.

### PR DESCRIPTION
**Pull request by rolfyone**
Accepts an array of public key strings to come back in json format
```
[
 "0x00", "0x11"
]
```

Adds keys to a set, which will de-duplicate any keys that are specified on both the command line and in the URL.

fixes #3378

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java [hunk 1]
```diff
@@ -173,4 +169,50 @@ private Path convertToPath(final String option) {
     }
     return Path.of(option);
   }
+
+  static class PublicKeyLoader {
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java]: nit: I'd probably move this to its own file (but still package private).

File: teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java [hunk 2]
```diff
@@ -173,4 +169,50 @@ private Path convertToPath(final String option) {
...
+        final Set<BLSPublicKey> blsPublicKeySet =
+            publicKeys.stream()
+                .filter(key -> !key.contains(":"))
+                .map(key -> BLSPublicKey.fromSSZBytes(Bytes.fromHexString(key)))
+                .collect(Collectors.toSet());
+        blsPublicKeySet.addAll(
+            publicKeys.stream()
+                .filter(key -> key.contains(":"))
+                .flatMap(this::readKeysFromUrl)
+                .collect(Collectors.toList()));
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java]: Probably just do this in one pass:
```suggestion
        final Set<BLSPublicKey> blsPublicKeySet =
            publicKeys.stream()
                .filter(key -> !key.contains(":"))
                .flatMap(value -> key.contains(':') ? readKeysFromUrl(value) : BLSPublicKey.fromSSZBytes(Bytes.fromHexString(value)))
                .collect(Collectors.toSet());
```

File: teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java [hunk 3]
```diff
@@ -173,4 +169,50 @@ private Path convertToPath(final String option) {
...
+      }
+    }
+
+    private Stream<BLSPublicKey> readKeysFromUrl(final String url) {
+      try {
+        final String[] keys = objectMapper.readValue(new URL(url), String[].class);
+        return Arrays.asList(keys).stream()
+            .map(key -> BLSPublicKey.fromSSZBytes(Bytes.fromHexString(key)));
+      } catch (IOException ex) {
+        throw new IllegalArgumentException("Failed to load public keys from URL", ex);
```
Comment (ajsutton) [inline teku/src/main/java/tech/pegasys/teku/cli/options/ValidatorKeysOptions.java]: Need to test that this provide a friendly error message - possibly needs to throw `InvalidConfigurationException` to make that happen.

**Review by ajsutton — APPROVED** [review]
LGTM.

**Comment by rolfyone** [comment]
for doc: this basically means example:
```
teku --eth1-deposit-contract-address=dddddddddddddddddddddddddddddddddddddddd \
--eth1-endpoint=http://localhost:8545 \
--validators-external-signer-public-keys=0xa99a...e44c,0xb89b...4a0b \
--validators-external-signer-url=http://localhost:9000
```

could be instead
```
teku --eth1-deposit-contract-address=dddddddddddddddddddddddddddddddddddddddd \
--eth1-endpoint=http://localhost:8545 \
--validators-external-signer-public-keys=http://localhost:9000/api/v1/eth2/publicKeys \
--validators-external-signer-url=http://localhost:9000
````

**Pull request merged**